nexus/db-queries/src/db/datastore/scim_provider_store.rs

```diff
@@ -7,124 +7,1791 @@
+        group_id: SiloGroupUuid,
+    ) -> Result<Option<IdOrdMap<GroupMember>>, diesel::result::Error> {
+        use nexus_db_schema::schema::silo_group_membership::dsl;
+
+        let users: Vec<Uuid> = dsl::silo_group_membership
+            .filter(dsl::silo_group_id.eq(to_db_typed_uuid(group_id)))
+            .select(dsl::silo_user_id)
+            .load_async(conn)
+            .await?;
+
+        if users.is_empty() {
+            Ok(None)
+        } else {
+            let members = users
+                .into_iter()
+                .map(|user_id| GroupMember {
+                    // Note neither the scim2-rs crate or Nexus support nested
+                    // groups
+                    resource_type: Some(ResourceType::User.to_string()),
+                    value: Some(user_id.to_string()),
+                })
+                .collect();
+
+            Ok(Some(members))
+        }
+    }
+
+    async fn get_user_by_id_in_txn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        _err: OptionalError<ProviderStoreError>,
+        user_id: SiloUserUuid,
+    ) -> Result<Option<StoredParts<User>>, diesel::result::Error> {
+        let maybe_user = {
+            use nexus_db_schema::schema::silo_user::dsl;
+
+            dsl::silo_user
+                .filter(dsl::silo_id.eq(self.authz_silo.id()))
+                .filter(
+                    dsl::user_provision_type.eq(model::UserProvisionType::Scim),
+                )
+                .filter(dsl::id.eq(to_db_typed_uuid(user_id)))
+                .filter(dsl::time_deleted.is_null())
+                .select(model::SiloUser::as_returning())
+                .first_async(conn)
+                .await
+                .optional()?
+        };
+
+        let Some(user) = maybe_user else {
+            return Ok(None);
+        };
+
+        let groups: Option<Vec<UserGroup>> =
+            self.get_user_groups_for_user_in_txn(conn, user_id).await?;
+
+        let SiloUser::Scim(user) = user.into() else {
+            // With the user provision type filter, this should never be another
+            // type.
+            unreachable!();
+        };
+
+        Ok(Some(convert_to_scim_user(user, groups)))
+    }
+
+    async fn create_user_in_txn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: OptionalError<ProviderStoreError>,
+        user_request: CreateUserRequest,
+    ) -> Result<StoredParts<User>, diesel::result::Error> {
+        use nexus_db_schema::schema::silo_user::dsl;
+
+        // userName is meant to be unique: If the user request is adding a
+        // userName that already exists, reject it
+
+        let maybe_other_user = dsl::silo_user
+            .filter(dsl::silo_id.eq(self.authz_silo.id()))
+            .filter(dsl::user_provision_type.eq(model::UserProvisionType::Scim))
+            .filter(lower(dsl::user_name).eq(lower(user_request.name.clone())))
```

Comment:
Because the DB itself isn't enforcing case-sensitive uniqueness, is it possible that two concurrent requests could write different cases of the same name at the same time? You might want a unique index like `CREATE UNIQUE INDEX ON silo_user (silo_id, LOWER(user_name)) WHERE user_provision_type = 'scim' AND time_deleted IS NULL`.

Reply:
I don't think two concurrent requests could race in this way because of the other calls to `lower` in the filters but a belt and suspenders approach here is the right one.

done in [6285129](https://github.com/oxidecomputer/omicron/pull/9276/commits/62851293e3cdc5d5fe0ffc962872711cc378b8ad)